modules/darwinosl/darwinosl-source.m

```diff
@@ -0,0 +1,559 @@
+{
+  darwinosl_source_persist_load_position(self->log_source_persist, &self->log_source_position);
+
+  /* NOTE: Here could be some internal logic to solve clashes might be caused by bookmarks saved with different bookmarking options, but
+           based on read-old-records() we just let the user manually solve this (e.g. clear the persis-file if needed)
+  */
+  NSDate *maxBookmarkDistanceDate = [NSDate dateWithTimeIntervalSinceNow:-1 * (NSTimeInterval)(
+                                              self->options.max_bookmark_distance)];
+  if (self->options.max_bookmark_distance > 0)
+    *startDate = maxBookmarkDistanceDate;
+
+  gboolean should_start_from_now = (FALSE == darwinosl_sd_options_get_read_old_records(&self->options) &&
+                                    (self->log_source_position.log_position == 0.0 || self->options.do_not_use_bookmark));
+  if ((FALSE == self->options.do_not_use_bookmark && self->log_source_position.log_position != 0.0)
+      || should_start_from_now)
+    {
+      NSDate *savedDate = [NSDate dateWithTimeIntervalSince1970:self->log_source_position.log_position];
+      *startDate = (should_start_from_now ?
+                    [NSDate dateWithTimeIntervalSinceNow:-1] :
+                    (self->options.max_bookmark_distance > 0 ? [maxBookmarkDistanceDate laterDate:savedDate] : savedDate)
+                   );
+      return (FALSE == should_start_from_now);
+    }
+  return FALSE;
+}
+
+static gsize
+_log_message_from_string(const char *msg_cstring, MsgFormatOptions *format_options, LogMessage **out_msg)
+{
+  gsize msg_len = strlen(msg_cstring);
+  *out_msg = msg_format_construct_message(format_options, (const guchar *) msg_cstring, msg_len);
+  msg_format_parse_into(format_options, *out_msg, (const guchar *) msg_cstring, msg_len);
+  return msg_len;
+}
+
+static gboolean
+_open_osl(LogThreadedSourceDriver *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+
+  self->curr_fetch_in_run = 0;
+
+  if ([self->osLogSource openStore] == nil)
+    return FALSE;
+
+  NSDate *startDate = nil;
+  bool positionRestored = _log_position_date_from_persist(self, &startDate);
+  msg_trace("darwinosl: Should start from",
+            evt_tag_str("position",
+                        (startDate ?
+                         [OSLogSource.RFC3339DateFormatter stringFromDateWithMicroseconds:startDate].UTF8String :
+                         "no saved position")));
+
+  if ([self->osLogSource openEnumeratorWithDate:startDate
+                         filterString:(self->options.filter_predicate ? [NSString stringWithUTF8String:self->options.filter_predicate] : nil)
+                         options:(self->options.go_reverse ? OSLogEnumeratorReverse : 0)] == nil)
+    {
+      [self->osLogSource closeAll];
+      return FALSE;
+    }
+
+  if (positionRestored)
+    {
+      OSLogEntry *nextLogEntry = [self->osLogSource fetchNextEntry];
```

Comment:
We are fetching a log entry after restoring our position, but is that  log entry delivered at the end? My assumption from reading the code is that fetchNextEntry() would move the current position, so by the time _fetch() is called, we would skip this one, unless we make sure that it gets delivered somehow.

But I don't see that here.

Reply:
The name `fetchNextEntry` is confusing a bit here unfortunately (I just kept it to follow the original Apple API name convention) :S

At the first time invocation the underlying NSEnumerator will fetch the first element from the collection that in this case is the one that the call of _open_osl (more precisely inside that the `entriesEnumeratorWithOptions:position:predicate:error` function) positioned the enumerator to (or `nil` if no such item anymore) that is the one we saved as the bookmark.
So basically it positions the enumerator "before" the bookmarked items pos.

This is why the _check_restored_postion in the following if() could work as well.